### PR TITLE
Edge case for linear filter and unit tests.

### PR DESCRIPTION
**Pull request by cameron-devine**

### CODE REVIEW

File: src/TimeSeriesFilter.cpp [hunk 1]
```diff
@@ -400,7 +400,14 @@ time_t TimeSeriesFilter::timeBefore(time_t t) {
     return 0;
   }
   else if (this->clock()) {
-    return this->clock()->timeBefore(t);
+    auto b4 = this->clock()->timeBefore(t);
+    if (_resampleMode == ResampleModeLinear && source()->timeAfter(b4) == 0) {
```
Comment (samhatchett) [inline src/TimeSeriesFilter.cpp]: one weird edge case on this. 
consider a source series with times: `[100,200,300]`
... then take a resampler with clock (100s), and ask for `timeBefore(400)`
The resampler (`this`) will go to its clock, get time before 400, which is 300.
Then this new logic fires. Linear mode true, and `source()->timeAfter(300)` will be zero, since there are no values after t=300. But! we should be returning a point at 300, since resampler and source align perfectly at that time.

I think the existing convention for this sort of situation is to back up by 1s. 
so something like:

```
if (_resampleMode == ResampleModeLinear && source()->timeAfter(b4 - 1) == 0) {
```

File: test/test_filters.cpp [hunk 2]
```diff
@@ -516,6 +516,65 @@ BOOST_AUTO_TEST_CASE(complex_resampler) {
...
+
+  auto stepRs = step->pointAtOrBefore(1643617189 + ONE_DAY);
+  BOOST_CHECK_EQUAL(stepRs.isValid, true);
+  BOOST_CHECK_EQUAL(stepRs.time, rs->clock()->timeBefore(1643617189 + ONE_DAY));
+  BOOST_CHECK_EQUAL(stepRs.value, 14.725202560424805);
+
+  auto stepRs2 = step->pointAtOrBefore(1643617189);
+  BOOST_CHECK_EQUAL(stepRs2.isValid, true);
+  BOOST_CHECK_EQUAL(stepRs2.time, rs->clock()->timeBefore(1643617189) );
+  BOOST_CHECK_EQUAL(stepRs2.value, raw->pointAtOrBefore(rs->clock()->timeBefore(1643617189)).value);
```
Comment (samhatchett) [inline test/test_filters.cpp:573]: these tests should be updated as per the comment above, respecting the difference between "before" and "atOrBefore"

**Review by samhatchett — CHANGES_REQUESTED** [review]

**Review by samhatchett — APPROVED** [review]
